File: apps/photokit/photokit/rename.py
#!/usr/bin/env python3
"""
Rename photos to a normalized date-based filename.

Can be run stand-alone:

    python -m photokit.rename /path/to/photos --dry-run

Or invoked as ``photokit rename`` via the Click CLI.
"""

import argparse
import sys
from collections import defaultdict
from pathlib import Path
from typing import Optional

from .date_patterns import (
    auto_detect,
    match_from_pattern,
    DATE_PATTERN_HELP,
)

IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".tiff", ".tif", ".heic", ".heif", ".webp"}


DEFAULT_TEMPLATE = "{year}-{month:02d}-{day:02d}_{hour:02d}-{minute:02d}-{second:02d}"


def find_images(directory: Path, extensions: set, recursive: bool) -> list[Path]:
    """Find image files matching *extensions* under *directory*."""
    if recursive:
        files = [p for ext in extensions for p in directory.rglob(f"*{ext}")]
    else:
        files = [p for ext in extensions for p in directory.glob(f"*{ext}")]
    return sorted(
        p for p in files
        if p.is_file() and not p.is_symlink() and not p.name.startswith("._")
    )


def render_template(template: str, parts: dict, seq: Optional[int] = None, ext: str = "") -> str:
    """Render a filename template from date parts.

    Supported placeholders: ``{year}``, ``{month}``, ``{day}``,
    ``{hour}``, ``{minute}``, ``{second}``, ``{seq}``, ``{ext}``.

    All date placeholders support zero-padded format specs
    (e.g. ``{month:02d}``).
    """
    ctx = {k: int(v) if isinstance(v, str) else v for k, v in parts.items()}
    ctx["ext"] = ext
    ctx["seq"] = seq or 0
    return template.format(**ctx)


def rename_photos(
    directory: Path,
    *,
    pattern: Optional[str] = None,
    template: str = DEFAULT_TEMPLATE,
    collision: str = "increment",
    recursive: bool = False,
    dry_run: bool = False,
    extensions: Optional[set] = None,
) -> int:
    """Main logic for rename.

    Returns exit code (0 = success).
    """
    if extensions is None:
        extensions = IMAGE_EXTENSIONS

    images = find_images(directory, extensions, recursive)
    if not images:
        print(f"No images found in {directory}")
        return 0

    print(f"\n📷  Found {len(images)} image(s) in {directory}\n")

    # Pre-scan: count how many files share each timestamp key
    # so we know whether to include a sequence number.
    timestamp_counts: dict[str, int] = defaultdict(int)
    date_parts_by_path: dict[Path, Optional[dict]] = {}

    for filepath in images:
        if pattern:
            parts = match_from_pattern(pattern, filepath.name)
        else:
            parts = auto_detect(filepath.name)

        date_parts_by_path[filepath] = parts
        if parts:
            key = render_template(template, parts, ext=filepath.suffix.lower())
            timestamp_counts[key] += 1

    renamed = 0
    unchanged = 0
    skipped = 0
    errors = 0
    seq_counter: dict[str, int] = defaultdict(int)

    for filepath in images:
        parts = date_parts_by_path[filepath]
        if parts is None:
            print(f"  ⚠️  Could not parse date from: {filepath.name}")
            skipped += 1
            continue

        ext = filepath.suffix.lower()
        base_key = render_template(template, parts, ext=ext)
        total_for_key = timestamp_counts[base_key]
        seq_counter[base_key] += 1
        seq = seq_counter[base_key]

        if total_for_key > 1:
            new_name = render_template(template, parts, seq=seq, ext=ext)
        else:
            new_name = render_template(template, parts, ext=ext)

        new_path = directory / new_name

        if new_path == filepath:
            # Already correctly named
            unchanged += 1
            continue

        # Handle collisions
        if new_path.exists() and new_path != filepath:
            if collision == "skip":
                print(f"  ⚠️  Collision — skipping: {filepath.name}")
                skipped += 1
                continue
            elif collision == "overwrite":
                pass  # will overwrite below
            else:  # increment
                # Append a hash suffix to disambiguate
                safe_base = new_path.stem
                new_name = f"{safe_base}_{seq:04d}{ext}"
                new_path = directory / new_name
                # If it *still* exists, add a hash of the original name
                while new_path.exists() and new_path != filepath:
                    new_name = f"{safe_base}_{seq:04d}_{hash(filepath.name) % 10000:04d}{ext}"
                    new_path = directory / new_name
                    seq += 1

        if dry_run:
            display_old = filepath.name[:60]
            display_new = new_name[:60]
            print(f"  🔄 {display_old}  →  {display_new}")
            renamed += 1
            continue

        try:
            filepath.rename(new_path)
            print(f"  ✅ {filepath.name[:60]}  →  {new_name[:60]}")
            renamed += 1
        except Exception as exc:
            print(f"  ❌ Error renaming {filepath.name}: {exc}", file=sys.stderr)
            errors += 1

    print(f"\n{'=' * 50}")
    print(f"  Renamed:   {renamed}")
    print(f"  Unchanged: {unchanged}")
    print(f"  Skipped:   {skipped}")
    print(f"  Errors:    {errors}")
    return 1 if errors else 0


# ---------------------------------------------------------------------------
# Stand-alone entry point
# ---------------------------------------------------------------------------

def main(argv: Optional[list[str]] = None) -> int:
    parser = argparse.ArgumentParser(
        description="Rename photos to a normalized date-based filename.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=DATE_PATTERN_HELP,
    )
    parser.add_argument("directory", type=Path, help="Directory containing photos")
    parser.add_argument(
        "-p", "--pattern",
        help="Custom regex with named groups year, month, day (or year, doy)",
    )
    parser.add_argument(
        "-t", "--template",
        default=DEFAULT_TEMPLATE,
        help=f"Output filename template (default: {DEFAULT_TEMPLATE})",
    )
    parser.add_argument(
        "--collision",
        default="increment",
        choices=["increment", "overwrite", "skip"],
        help="How to handle name collisions (default: increment)",
    )
    parser.add_argument(
        "-r", "--recursive", action="store_true",
        help="Recurse into subdirectories",
    )
    parser.add_argument(
        "-n", "--dry-run", action="store_true",
        help="Preview only — don't rename",
    )
    parser.add_argument(
        "--extensions",
        default=",".join(sorted(IMAGE_EXTENSIONS)),
        help=f"Comma-separated extensions (default: {','.join(sorted(IMAGE_EXTENSIONS))})",
    )

    args = parser.parse_args(argv)
    extensions = {f".{e.strip().lstrip('.')}" for e in args.extensions.split(",")}

    return rename_photos(
        args.directory,
        pattern=args.pattern,
        template=args.template,
        collision=args.collision,
        recursive=args.recursive,
        dry_run=args.dry_run,
        extensions=extensions,
    )


if __name__ == "__main__":
    sys.exit(main())

File: apps/imagekit/src/commands/rename.ts
import fs from "node:fs";
import path from "node:path";
import { autoDetect, matchFromPattern } from "../date-patterns";

interface RenameOptions {
  pattern?: string;
  template: string;
  collision: string;
  recursive: boolean;
  dryRun: boolean;
  extensions?: string;
}

const DEFAULT_TEMPLATE = "{year}-{month:02d}-{day:02d}_{hour:02d}-{minute:02d}-{second:02d}";

const DEFAULT_EXTENSIONS = new Set([
  ".jpg",
  ".jpeg",
  ".png",
  ".tiff",
  ".tif",
  ".heic",
  ".heif",
  ".webp",
]);

function parseExtensions(extStr?: string): Set<string> {
  if (!extStr) return DEFAULT_EXTENSIONS;
  const set = new Set<string>();
  for (const e of extStr.split(",")) {
    const trimmed = e.trim().toLowerCase();
    if (!trimmed) continue;
    const withDot = trimmed.startsWith(".") ? trimmed : `.${trimmed}`;
    set.add(withDot);
  }
  return set.size > 0 ? set : DEFAULT_EXTENSIONS;
}

function findImages(directory: string, extensions: Set<string>, recursive: boolean): string[] {
  const results: string[] = [];
  const dir = path.resolve(directory);

  function walk(current: string) {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(current, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.name.startsWith("._")) continue;
      const full = path.join(current, entry.name);
      if (entry.isSymbolicLink()) continue;
      if (entry.isDirectory()) {
        if (recursive) walk(full);
      } else if (entry.isFile()) {
        const ext = path.extname(entry.name).toLowerCase();
        if (extensions.has(ext)) results.push(full);
      }
    }
  }

  walk(dir);
  return results.sort();
}

function renderTemplate(template: string, parts: Record<string, number | string>, seq?: number | null, ext?: string): string {
  // Support Python-style format specs: {year}, {month:02d}, {day:02d}, etc.
  // We translate {key:02d} -> zero-padded number
  const ctx: Record<string, unknown> = { ...parts };
  ctx["ext"] = ext ?? "";
  ctx["seq"] = seq ?? 0;

  // Replace each {key[:format]} with value
  // Supported format: 02d, 2d, d
  return template.replace(/\{(\w+)(?::([^}]+))?\}/g, (_match, key: string, fmt: string | undefined) => {
    const val = ctx[key];
    if (val === undefined) {
      throw new Error(`Unknown template placeholder: ${key}`);
    }
    if (fmt) {
      // Handle integer formatting like 02d, 04d
      const m = fmt.match(/^0?(\d+)d$/);
      if (m) {
        const width = parseInt(m[1]!, 10);
        const num = typeof val === "number" ? val : parseInt(String(val), 10);
        if (Number.isNaN(num)) return String(val);
        return String(num).padStart(width, "0");
      }
      // Fallback: just string
      return String(val);
    }
    return String(val);
  });
}

export function cmdRename(directory: string, opts: RenameOptions): void {
  const extensions = parseExtensions(opts.extensions);
  const dir = path.resolve(directory ?? ".");
  if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) {
    console.error(`Error: directory not found: ${directory}`);
    process.exit(1);
  }

  const template = opts.template ?? DEFAULT_TEMPLATE;
  const collision = opts.collision ?? "increment";

  const images = findImages(dir, extensions, !!opts.recursive);
  if (images.length === 0) {
    console.log(`No images found in ${dir}`);
    return;
  }

  console.log(`\n📷  Found ${images.length} image(s) in ${dir}\n`);

  // Pre-scan to count collisions for template without seq
  const timestampCounts = new Map<string, number>();
  const datePartsByPath = new Map<string, (ReturnType<typeof autoDetect>)>();

  for (const filepath of images) {
    const name = path.basename(filepath);
    let parts: ReturnType<typeof autoDetect> = null;
    try {
      if (opts.pattern) {
        parts = matchFromPattern(opts.pattern, name);
      } else {
        parts = autoDetect(name);
      }
    } catch (e) {
      console.error(`  ❌ Invalid pattern: ${e instanceof Error ? e.message : String(e)}`);
      process.exit(1);
    }
    datePartsByPath.set(filepath, parts);
    if (parts) {
      const ext = path.extname(filepath).toLowerCase();
      const key = renderTemplate(template, parts as unknown as Record<string, number>, null, ext);
      // The key includes extension? In python, they included ext when counting
      // python: key = render_template(template, parts, ext=filepath.suffix.lower())
      // That extension will be interpolated only if template contains {ext}
      // Otherwise ext not in key but they still used it; We'll follow same: if template contains {ext}, then key distinct per ext
      // Otherwise same key regardless of ext will still count
      timestampCounts.set(key, (timestampCounts.get(key) ?? 0) + 1);
    }
  }

  let renamed = 0;
  let unchanged = 0;
  let skipped = 0;
  let errors = 0;
  const seqCounter = new Map<string, number>();

  for (const filepath of images) {
    const parts = datePartsByPath.get(filepath);
    if (!parts) {
      console.log(`  ⚠️  Could not parse date from: ${path.basename(filepath)}`);
      skipped++;
      continue;
    }

    const ext = path.extname(filepath).toLowerCase();
    const baseKey = renderTemplate(template, parts as unknown as Record<string, number>, null, ext);
    const totalForKey = timestampCounts.get(baseKey) ?? 0;
    const seqCount = (seqCounter.get(baseKey) ?? 0) + 1;
    seqCounter.set(baseKey, seqCount);
    const seq = seqCount;

    let newName: string;
    if (totalForKey > 1) {
      newName = renderTemplate(template, parts as unknown as Record<string, number>, seq, ext);
      // If template doesn't contain seq, need to ensure uniqueness anyway
      if (newName === baseKey && seq > 1) {
        // Append seq manually if collision would remain
        const stem = path.basename(newName, ext);
        newName = `${stem}_${String(seq).padStart(4, "0")}${ext}`;
      }
    } else {
      newName = renderTemplate(template, parts as unknown as Record<string, number>, null, ext);
      // If template lacks {ext}, add ext
      if (!template.includes("{ext}") && !newName.toLowerCase().endsWith(ext)) {
        newName = `${newName}${ext}`;
      }
    }

    // Ensure newName has extension if template didn't include it
    if (!path.extname(newName)) {
      newName = `${newName}${ext}`;
    }

    const newPath = path.join(dir, newName);

    if (path.resolve(newPath) === path.resolve(filepath)) {
      unchanged++;
      continue;
    }

    if (fs.existsSync(newPath) && path.resolve(newPath) !== path.resolve(filepath)) {
      if (collision === "skip") {
        console.log(`  ⚠️  Collision — skipping: ${path.basename(filepath)}`);
        skipped++;
        continue;
      } else if (collision === "overwrite") {
        // allow overwrite
      } else {
        // increment
        const safeBase = path.basename(newName, ext);
        let candidate = `${safeBase}_${String(seq).padStart(4, "0")}${ext}`;
        let candidatePath = path.join(dir, candidate);
        let extraSeq = seq;
        while (fs.existsSync(candidatePath) && path.resolve(candidatePath) !== path.resolve(filepath)) {
          // Append hash of original name modulo 10000 as python did
          let hash = 0;
          for (let i = 0; i < filepath.length; i++) hash = (hash * 31 + filepath.charCodeAt(i)) % 10000;
          candidate = `${safeBase}_${String(extraSeq).padStart(4, "0")}_${String(hash).padStart(4, "0")}${ext}`;
          candidatePath = path.join(dir, candidate);
          extraSeq++;
          if (extraSeq > seq + 10) break;
        }
        newName = candidate;
      }
    }

    const finalPath = path.join(dir, newName);

    if (opts.dryRun) {
      const displayOld = path.basename(filepath).slice(0, 60);
      const displayNew = newName.slice(0, 60);
      console.log(`  🔄 ${displayOld}  →  ${displayNew}`);
      renamed++;
      continue;
    }

    try {
      fs.renameSync(filepath, finalPath);
      console.log(`  ✅ ${path.basename(filepath).slice(0, 60)}  →  ${newName.slice(0, 60)}`);
      renamed++;
    } catch (exc) {
      console.error(`  ❌ Error renaming ${path.basename(filepath)}: ${exc instanceof Error ? exc.message : String(exc)}`);
      errors++;
    }
  }

  console.log(`\n${"=".repeat(50)}`);
  console.log(`  Renamed:   ${renamed}`);
  console.log(`  Unchanged: ${unchanged}`);
  console.log(`  Skipped:   ${skipped}`);
  console.log(`  Errors:    ${errors}`);
  if (errors > 0) process.exitCode = 1;
}
